src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -137,15 +137,32 @@ double LMNNFunction<MetricType>::Evaluate(const arma::mat& transformation)
         bool exactEval = true;
 
         // Bounds for eval.
-        if (transformationOld.n_elem != 0 && !std::isnan(evalOld[i](j, l)))
+        if (transformationOld.n_elem != 0 && !std::isnan(evalOld(j, l, i)))
         {
-          eval = evalOld[i](j, l) + transformationDiff *
-            (norm(targetNeighbors(j, i)) + norm(impostors(l, i)) +
+          double maxNorm;
+          if (!std::isnan(maxImpNorm(l, i)))
+            maxNorm = std::max(norm(impostors(l, i)), maxImpNorm(l, i));
+          else
+            maxNorm = norm(impostors(l, i));
```

Comment:
We could avoid the `if` statement here if we initialize `maxImpNorm` elements to 0.  (Correct me if I am wrong---but I think that will work.)  You could even just do

```
maxImpNorm(l, i) = std::max(maxImpNorm(l, i), norm(impostors(l, i)));
```

in that case.

Reply:
Definitely, that's a better way of doing this.